backend/coins/coin/amount.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2018 Shift Devices AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package coin
+
+import (
+	"math/big"
+	"strings"
+
+	"github.com/digitalbitbox/bitbox-wallet-app/util/errp"
+)
+
+// Amount represents an amount in the smallest coin unit (e.g. satoshi).
+type Amount struct {
+	amount *big.Int
+}
+
+// NewAmount creates a new amount.
+func NewAmount(amount *big.Int) Amount {
+	return Amount{amount: amount}
+}
+
+// NewAmountFromInt64 creates a new amount.
+func NewAmountFromInt64(amount int64) Amount {
+	return Amount{amount: big.NewInt(amount)}
+}
+
+// NewAmountFromString parses a user given coin amount, converting it from the default coin unit to
+// the the smallest unit.
+func NewAmountFromString(s string, unit *big.Int) (Amount, error) {
+	// big.Rat parsing accepts rationals like "2/3". Exclude those, we only want decimals.
+	if strings.ContainsRune(s, '/') {
+		return Amount{}, errp.Newf("could not parse %q", s)
+	}
+	rat, ok := new(big.Rat).SetString(s)
+	if !ok {
+		return Amount{}, errp.Newf("could not parse %q", s)
+	}
+	rat.Mul(rat, new(big.Rat).SetInt(unit))
+	if rat.Denom().Cmp(big.NewInt(1)) != 0 {
+		return Amount{}, errp.Newf("invalid amount %q", s)
+	}
+	return Amount{amount: rat.Num()}, nil
+}
+
+// Int64 returns the int64 representation of amount.
+// If x cannot be represented in an int64, the result is undefined.
+func (amount *Amount) Int64() int64 {
+	return amount.amount.Int64()
+}
+
+// Int returns a copy of the underlying big integer.
+func (amount *Amount) Int() *big.Int {
```

Comment:
I would try and avoid exposing big int due to the same concerns - it's mutable.

Reply:
ah sorry, it returns a copy. still, helper methods like above for "negative" and "zero" might be a more readable solition?

Reply:
I guess, but on the other hand, it's the stdlib, so people should be able to use it. Not sure if it's a good idea to wrap every underlying functionality I need instead of just giving access.

For example, in other coins, big.Int could be the basis for calculations, and wrapping all the arithmetic functions is overkill?

Reply:
Yeah, that's a downside. I haven't checked but are you really doing some math with the amount? I thought this was just to parse user input and use the value as is, no arithmetic operations.

Either way, I'd suggest renaming the method to `BigInt`. You already have `Int64` which returns `int64` so first thing that comes in mind is `Int()` returns the `int` type unless I go and look up the signature.

Reply:
I use it when formatting the amount (could be a struct method too), and Ethereum will use big.Int internally, so arithmetics will be done there (i.e. adding/subtracting fees, etc.).